vm/src/ser_de.rs

```diff
@@ -0,0 +1,199 @@
+            let mut map = serializer.serialize_map(Some(pairs.len()))?;
+            for (key, e) in pairs.iter() {
+                map.serialize_entry(&self.clone_with_object(key), &self.clone_with_object(&e))?;
+            }
+            map.end()
+        } else if self.pyobject.is(&self.vm.get_none()) {
+            serializer.serialize_none()
+        } else {
+            Err(serde::ser::Error::custom(format!(
+                "Object of type '{:?}' is not serializable",
+                self.pyobject.class()
+            )))
+        }
+    }
+}
+
+// This object is used as the seed for deserialization so we have access to the PyContext for type
+// creation
+#[derive(Clone)]
+pub struct PyObjectDeserializer<'c> {
+    vm: &'c VirtualMachine,
+}
+
+impl<'c> PyObjectDeserializer<'c> {
+    pub fn new(vm: &'c VirtualMachine) -> Self {
+        PyObjectDeserializer { vm }
+    }
+}
+
+impl<'de> serde::de::DeserializeSeed<'de> for PyObjectDeserializer<'de> {
+    type Value = PyObjectRef;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_any(self.clone())
+    }
+}
+
+impl<'de> Visitor<'de> for PyObjectDeserializer<'de> {
+    type Value = PyObjectRef;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a type that can deserialise in Python")
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(self.vm.ctx.new_str(value.to_string()))
+    }
+
+    fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(self.vm.ctx.new_str(value))
+    }
+
+    fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        // The JSON deserializer always uses the i64/u64 deserializers, so we only need to
+        // implement those for now
+        Ok(self.vm.ctx.new_int(value))
+    }
+
+    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        // The JSON deserializer always uses the i64/u64 deserializers, so we only need to
+        // implement those for now
+        Ok(self.vm.ctx.new_int(value))
+    }
+
+    fn visit_f64<E>(self, value: f64) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(self.vm.ctx.new_float(value))
+    }
+
+    fn visit_bool<E>(self, value: bool) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(self.vm.ctx.new_bool(value))
+    }
+
+    fn visit_seq<A>(self, mut access: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::SeqAccess<'de>,
+    {
+        let mut seq = Vec::with_capacity(access.size_hint().unwrap_or(0));
+        while let Some(value) = access.next_element_seed(self.clone())? {
```

Comment:
It's probably a later topic, since you just copy pasted this code from json.rs, but what happens with a nested data structure like this?
```python
L = []
L.append(L)
serialize(L) --> BOOM?
```

Reply:
Right now, a segfault. CPython errors with a `ValueError: Circular reference detected`.